logfire/_internal/integrations/llm_providers/openai.py

```diff
@@ -72,10 +70,42 @@ def content_from_completions(chunk: Completion | None) -> str | None:
     return None  # pragma: no cover
 
 
-def content_from_chat_completions(chunk: ChatCompletionChunk | None) -> str | None:
-    if chunk and chunk.choices:
-        return chunk.choices[0].delta.content
-    return None
+class OpenaiCompletionStreamState(StreamState):
+    def __init__(self):
+        self._content: list[str] = []
+
+    def record_chunk(self, chunk: Completion) -> None:
+        content = content_from_completions(chunk)
+        if content:
+            self._content.append(content)
+
+    def get_response_data(self) -> Any:
+        return {'combined_chunk_content': ''.join(self._content), 'chunk_count': len(self._content)}
+
+
+class OpenaiChatCompletionStreamState(StreamState):
+    def __init__(self):
+        self._stream_state = ChatCompletionStreamState(
+            # We do not need the response to be parsed into Python objects so can skip
+            # providing the `response_format` and `input_tools` arguments.
+            input_tools=openai.NOT_GIVEN,
+            response_format=openai.NOT_GIVEN,
+        )
+
+    def record_chunk(self, chunk: ChatCompletionChunk) -> None:
+        self._stream_state.handle_chunk(chunk)
+
+    def get_response_data(self) -> Any:
+        try:
+            final_completion = self._stream_state.current_completion_snapshot
+        except AssertionError:
+            # AssertionError is raised when there is no completion snapshot
+            # Return empty content to show an empty Assistant response in the UI
+            return {'combined_chunk_content': ''}
+        return {
+            'message': final_completion.choices[0].message if final_completion.choices else None,
+            'usage': final_completion.usage,
```

Comment:
I worry about this having a significantly different shape from the other response data dicts.

Reply:
ah, i didn't realise we were already using this shape. yes, creating some types sounds good. @dmontagu any thoughts? should we consider these shapes stable?